### ISSUE TITLE
[apex] New Rules ideas

### ISSUE THREAD

**Issue by rsoesemann** [pmd/pmd#2322]
I collected ideas for new rules in the community and need to collect them first before I can detail them in single issues:

### Daniel Ballinger @FishOfPrey

- null checking for Booleans, integers, etc... If they are used in if conditions it would be common to do something like if(someBoolean) {} and miss the null case that Apex permits.
- Shadowing any system namespace or classes
- Variables that are only ever assigned but never otherwise used
- Putting multiple statements on the single line. This is less than ideal for code coverage calculations.
- People using =: as an operator in Apex SOQL variable binding. It leads to confusion. The : should be associated with the variable being bound. Not the =.

### Diéffrei Quadros @dieffrei

- Avoid creating multiple triggers on the same object
- Avoid leaving TODO/FIXME comments in code
- Avoid Literals in If Condition

**Comment by rsoesemann** [comment]
@gwilymatgearset @noerremark I asked a few people in the community which rules they would like to have in PMD for Apex and this is the result. 

In case you like one and think about contributing them lets create a real issue where we also describe acceptance tests and descriptions. I am happy to help. 

My favorites are (because of their usefulness and their easiness to implement them:

- Only one return per method XPATH
- Field Declarations Should be at the Start of a Class
- Configurable Brace Positioning 
- @AuraEnabled should   have Proper Error Handling
- Avoid Losing   Exception Information

**Comment by noerremark** [comment]
@rsoesemann - I like some of them, but there are also some that would need a `Controversial`-label or something like one return per method. But anyway, I'm a rule-o-holic :-)

I think there are some good things in [Apex Testing Best Practices](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_testing_best_practices.htm) too, like:

- Use the `runAs` method to test your application in different user contexts.
- Create all test data before calling the Test.startTest method.
- Use the ORDER BY keywords to ensure that the records are returned in the expected order.

> Configurable Brace Positioning
Is that a PMD-thing these days? I would love it since coverage is per line, not conditionals.

**Comment by noerremark** [comment]
> Avoid creating multiple triggers on the same object

I like that, but can PMD scan multiple files against each other these days?

**Comment by rsoesemann** [comment]
@gwilymatgearset as you already implemented the other complexity rule "easy as Sunday morning" would you be interested to also implement the "Bumby road rule" https://empear.com/blog/bumpy-road-code-complexity-in-context/?

**Comment by rsoesemann** [comment]
@noerremark you are right Multifile analysis is not working in PMD so this is not a rule we can make.

**Comment by gwilymatgearset** [comment]
> @gwilymatgearset as you already implemented the other complexity rule "easy as Sunday morning" would you be interested to also implement the "Bumby road rule" https://empear.com/blog/bumpy-road-code-complexity-in-context/?

I feel like you start getting diminishing returns as you add more complexity metrics :). I'm happy to do a bit of work on some of the rules. The ones I like the sound of most are:

- Variables that are only ever assigned but never otherwise used
- Field Declarations Should be at the Start of a Class
- Configurable Brace Positioning (where braces go is a hot topic at Gearset :stuck_out_tongue:)

I think that'll give me a nice grounding across various parts of the PMD code base so I'll be able to do bigger bits of work in the future. What do you think?

**Comment by armudgal** [comment]
If we can have something like Java has for [BrokenNullCheck](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#brokennullcheck), it would be great

**Comment by daveespo** [comment]
@rsoesemann  -- I want to +1 the Variable Shadowing item in the description on this issue. Not only should it protect against shadowing of system classes, objects and keywords, it should protect against shadowing of other classes/objects within the project

We could probably get by with a poor-mans first step by just protecting against the common error of creating a variable with the same name as the object it's an instance of .. too many Java devs go to this style:

```
Contact contact = new Contact();
contact.FirstName = 'Fred';
...
```

**Comment by rsoesemann** [comment]
@daveespo i also call my variables like classes and never ran into issues. Actually I see it as an antipattern if younger devs invent terrible alternatives like 

Contact contactRecord = 

just to not say

Contact contact =

Can you please elaborate why you and others see this as an issue?

**Comment by daveespo** [comment]
If you define a new class like so:

```
public with sharing class SomeClass {
	public void sayHello(){
		System.debug('Hello');
	}
}
```

And then you start using it throughout your project like this:

```
SomeClass someClass = new SomeClass();
someClass.sayHello();
```

Then, some smart developer comes along and adds a new `static` method

```
	public static void sayGoodbye(){
		System.debug('Hello');
	}
```

You find that you need to go through and rename all of your variables within each scope where they're declared otherwise you can't compile this:

```
SomeClass someClass = new SomeClass();
someClass.sayHello();
SomeClass.sayGoodbye();

Compile failure on line 3, column 11: Static method cannot be referenced from a non static context
```

This isn't an egregious problem to solve, and luckily the compiler tells you about it. Still, depending on the complexity of the project, the 'blast radius' of the fixup might be large.

Another [WTFApex ](https://github.com/ChuckJonas/wtfapex#shadowing-system-global-classes) moment is this example. Is `contact` a class or an instance of a `String`? Apparently it's both?

```
public with sharing class SomeClass {

	private String contact;

	public SomeClass(){
		contact = 'Fred Jones';
	}

	public void helloWorld(){
		contact bob = new contact();
		contact = 'Which is it?';
	}
}
```

**Comment by rsoesemann** [comment]
Ok and how should that be prevented?! By forbidding developer to call a contact instance contact ? I would not want such a rule so you will have to contribute it yourself 😉

**Comment by daveespo** [comment]
:-) I'm just trying to figure out which version of the rule about "Shadowing any system namespace or classes" you were referring to when you opened this request .. and for sure, I agree with @noerremark -- some of these are `Controversial` ;-)

**Comment by rsoesemann** [comment]
Not sure but I was just collecting a few ideas from my network and this ticket was a post it to memorize it. I think Daniel only intended to warn users when I overwrite System classes like Test which magically makes a lot of my code fail.
